File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        
        maps = [{} for x in range(len(nums))]
        result=0
        N = len(nums)
        
        for i in range(1, N):
            for j in range(0, i):
                cd = nums[i]-nums[j]
                
                countAtJ = maps[j].get(cd, 0)
                countAtI = maps[i].get(cd, 0)
                
                result += countAtJ
                maps[i][cd] = countAtI + countAtJ + 1
        
        return result
    

File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.java
class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        
        int n = nums.length;
        Map<Long, Integer>[] maps = new HashMap[n];
        for(int i=0; i<n; i++){
            maps[i] = new HashMap<>();
        }
        
        int result=0;
        for(int i=0; i<n; i++){
            for(int j=0; j<i; j++){
                
                long cd = (long)nums[i] - (long)nums[j];
                
                int countAtJ = maps[j].getOrDefault(cd, 0);
                int countAtI = maps[i].getOrDefault(cd, 0);
                
                maps[i].put(cd, countAtI + countAtJ + 1);
                
                result += countAtJ;
            }
        }
        
        return result;
    }
}